unified_ecommerce/test_utils.py

```diff
@@ -138,3 +223,211 @@ class PickleableMock(Mock):
+
+    @pytest.mark.parametrize("isLoggedIn", [True, False])
+    def test_list(self, isLoggedIn, client, user_client):
+        """
+        Test that the viewset can list objects.
+
+        Args:
+        - isLoggedIn (bool): whether or not the client is logged in
+        - client (APIClient): the client to use for non-logged in requests
+        - user_client (APIClient): the client to use for logged in requests
+        """
+
+        response = self._test_retrieval(
+            user_client if isLoggedIn else client,
+            self.list_url,
+            "list_url",
+            test_non_authenticated=not isLoggedIn,
+        )
+
+        if isLoggedIn:
+            assert "count" in response.data
+            assert response.data["count"] == self.queryset.count()
+
+    @pytest.mark.parametrize("isLoggedIn", [True, False])
+    def test_retrieve(self, isLoggedIn, client, user_client):
+        """
+        Test that the viewset can retrieve an object.
+
+        Args:
+        - isLoggedIn (bool): whether or not the client is logged in
+        - client (APIClient): the client to use for non-logged in requests
+        - user_client (APIClient): the client to use for logged in requests
+        """
+        instance = self.factory_class()
+        instance_qs = self.queryset.filter(pk=instance.pk)
+
+        response = self._test_retrieval(
+            user_client if isLoggedIn else client,
+            self.object_url.format(instance.pk),
+            "object_url",
+            test_non_authenticated=not isLoggedIn,
+        )
+
+        if isLoggedIn and instance.is_active is None:
+            dj_serializer = queryset_to_json(instance_qs)
+            assert_json_equal(
+                *make_timestamps_matchable([response.data, dj_serializer])
+            )
+
+        if isLoggedIn and instance.is_active is not None:
+            # Deleted items should return a 404. (is_active is None if it's active;
+            # otherwise it has a timestamp of when it was deleted.)
+            assert response.status_code == 404
+
+    def test_update(self, update_data, isLoggedIn, client, user_client):
+        """
+        Test that the viewset can update an object.
+
+        Args:
+        - update_data (dict): the data to use for the update
+        - isLoggedIn (bool): whether or not the client is logged in
+        - client (APIClient): the client to use for non-logged in requests
+        - user_client (APIClient): the client to use for logged in requests
+
+        Returns:
+        - tuple of (instance, response): the instance that was updated and the response
+        """
+        instance = self.factory_class()
+
+        use_client = user_client if isLoggedIn else client
+        response = use_client.patch(
+            self.object_url.format(instance.pk), data=update_data
+        )
+
+        if not isLoggedIn:
+            assert response.status_code == 403
```

Comment:
Should this assert a 200 in an `else` to be sure a 500 error isn't a passing test?

Reply:
For this (and the related comment on in `test_create`) I added an assertion to make sure the error code was under 500 - this way we can keep this simple and allow for tests where we're expecting a 400-level code back. (I test elsewhere where create passes in bad data and update tries to update a deleted record; I expect both of those to fail.)